### PR TITLE
Ayr 1142/non js render

### PR DESCRIPTION
**Pull request by colinbowen**
## Changes in this PR
- Add Support for rendering PDF & Image (PNG/JPG) on browsers that have JS disabled

## JIRA ticket
https://national-archives.atlassian.net/browse/AYR-1141

### After
Screenshot 2024-07-26 at 11 51 30

Screenshot 2024-07-26 at 11 51 45

Screenshot 2024-07-26 at 11 54 59
Screenshot 2024-07-26 at 11 54 46

- [ ] Requires env variable(s) to be updated

### CODE REVIEW

File: app/__init__.py
Comment (ltrasca) [inline app/__init__.py]: if there is no options around using these it would be good to figure out which ones we really need and which unsafe flags we can get rid of, also if we allow unsafe eval/inline scripts we should maybe look into sanitising any query strings we use - this could maybe be in a different ticket

File: app/static/src/scss/includes/_record.scss
Comment (ltrasca) [inline app/static/src/scss/includes/_record.scss]: any way we can have some of these classes under the same .uv class namespace? as in .uv { .mainPanel ... } just for the sake of it being slightly more organized

File: package.json
Comment (ltrasca) [inline package.json]: do we need uv as an npm dependency if we import it from a cdn?

**Review by ltrasca — COMMENTED** [review]
just a few minor comments from me, all else looks gud

File: package.json
Comment (colinbowen) [reply package.json]: Yes that's right, I've removed this.

File: app/static/src/scss/includes/_record.scss
Comment (colinbowen) [reply app/static/src/scss/includes/_record.scss]: Updated to move classes under .uv class

File: app/__init__.py
Comment (anthonyhashemi) [inline app/__init__.py]: is this allowing any scripts from these domains? Is it possible to lock it down to specific urls of those we need, or do they change dynamically?

File: app/__init__.py
Comment (anthonyhashemi) [inline app/__init__.py:122]: what is the purpose of this additional param. documentation, is it so we ignore worker-src and img-src? Either way, why did we need to treat them differently to the rest?

File: app/static/src/scss/includes/_record.scss
Comment (anthonyhashemi) [inline app/static/src/scss/includes/_record.scss:3]: I have skipped over this file as well as other css files, so worth getting someone else to review.

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py:736]: does this need to be a an endpoint if we only need the response in the `record` route? If not, we could just make it a function?

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py:865]: same with this, do we have a need for calling this as an api endpoint? If not, we can just make it a function

File: app/templates/main/record.html
Comment (anthonyhashemi) [inline app/templates/main/record.html]: as discussed, we're going to park html view for now, so this block can be removed

File: app/templates/main/record.html
Comment (anthonyhashemi) [inline app/templates/main/record.html]: could be worth extracting record details and record  view blocks into their own individual templates which we can include here

File: app/static/init.uv.js
Comment (anthonyhashemi) [inline app/static/init.uv.js:48]: why do we need to remove this element?

File: app/static/init.uv.js
Comment (anthonyhashemi) [inline app/static/init.uv.js:40]: could we decouple the tab switching logic from the javascript and instead just do form submissions on clicking of the tabs? If we do that, we should then be able to reuse the tabs for when javascript is disabled also like we discussed in the demo

File: app/main/routes.py
Comment (anthonyhashemi) [reply app/main/routes.py:736]: ah Im just piecing it all together now I think. Do you have documentation that explains the manifest and file url generation?

File: app/static/accessibility_audit_report.pdf
Comment (anthonyhashemi) [inline app/static/accessibility_audit_report.pdf:1]: what changed in the AA report pdf?

File: app/main/routes.py
Comment (anthonyhashemi) [reply app/main/routes.py:736]: I think it might be useful to chat through this, as Im wondering whether we should have a separate api for this manifest

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py]: is this different from the download route?

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py:736]: if we do need to keep the endpoint, would it be better to define the url as something like /record/id/manifest, like we have record/id/download etc as its not creating a new entry in db or anything like that, just buulding up a response for that record which will be consistent

File: app/static/init.uv.js
Comment (colinbowen) [reply app/static/init.uv.js:48]: Discussed on call. Also providing a screenshot of the attribution element which is removed. (Bottom left corner toolbar which contains the filename).

Screenshot 2024-08-05 at 12 05 04

File: app/static/src/scss/includes/_record.scss
Comment (colinbowen) [reply app/static/src/scss/includes/_record.scss:3]: Will request another review from Lucian.

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py]: Removed unsafe eval and limited script-src & style-src to specific universal viewer sources

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py:122]: Reviewed on call and updated while pairing.

File: app/static/accessibility_audit_report.pdf
Comment (colinbowen) [reply app/static/accessibility_audit_report.pdf:1]: Added pdf back. This should not have been deleted.

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py]: Reviewed on call and updated while pairing. Awaiting a review from AJ

File: app/main/routes.py
Comment (colinbowen) [reply app/main/routes.py:736]: This URL has been renamed to record/id/manifest. It is protected and does require authentication to be accessed.

File: app/main/routes.py
Comment (colinbowen) [reply app/main/routes.py]: Changed the download route to include the changes for render. An request arg is used to differentiate between needing to download the file as an attachment or just return the file for render

File: app/templates/main/record.html
Comment (colinbowen) [reply app/templates/main/record.html]: Removed any references to Record HTML view

File: app/templates/main/record-view.html
Comment (ahosgood) [inline app/templates/main/record-view.html]: Would it be better to include a SRI hash in this resource and not rely on jsdelivr's auto optimisation?

```suggestion
            <script nonce="{{ csp_nonce() }}" src="https://cdn.jsdelivr.net/npm/universalviewer@4.0.25/dist/umd/UV.js" integrity="sha256-Fdp0oui1Jul9OK4Kuwm7OSbhzuuqBcgD5wi8A2tXTns=" crossorigin="anonymous"></script>
```

There is some advice on this: https://nationalarchives.github.io/engineering-handbook/third-party/cdn/

File: app/static/init.uv.js
Comment (ahosgood) [inline app/static/init.uv.js:79]: What happens if the window is resized? Does this need to be reevaluated?

File: app/__init__.py
Comment (ahosgood) [inline app/__init__.py]: Is there any way to avoid using `unsafe-hashes` as it devalues the rest of the CSP? What events or code is being run that needs this?

File: app/static/init.uv.js
Comment (colinbowen) [reply app/static/init.uv.js:79]: Yes this will need to be updated for resized windows

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py]: UniversalViewer looks to pull some CSS which requires this, currently investigating

File: app/templates/main/record-view.html
Comment (anthonyhashemi) [reply app/templates/main/record-view.html]: I understand the benefit of adding out of interest but @ahosgood what do you mean by " not rely on jsdelivr's auto optimisation"?

File: app/__init__.py
Comment (anthonyhashemi) [reply app/__init__.py:122]: Can we just specify script-src since we only refer to `csp_nonce` in the script tags?

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py]: I get this was useful for development but do we want to keep this print?

File: app/templates/main/record.html
Comment (anthonyhashemi) [inline app/templates/main/record.html]: nice, much cleaner now these are extracted out 👌

File: app/main/routes.py
Comment (colinbowen) [reply app/main/routes.py]: Yep will remove this

File: app/templates/main/record-html-view.html
Comment (anthonyhashemi) [inline app/templates/main/record-html-view.html]: can get rid of this file as we dont need to support html view

File: app/static/uv.min.css
Comment (anthonyhashemi) [inline app/static/uv.min.css]: Ok so we went with serving the uv css file ourselves but leaving the js to coem from the external cdn.

Remind me, what benefit do we get fro serving this file ourself

File: app/templates/main/record-view.html
Comment (ahosgood) [reply app/templates/main/record-view.html]: Universal Viewer contains the "main" file `UV.js`.

jsDelivr has a feature that automatically minifies non-minified resources. By requesting `UV.min.js`, you are asking for the version minified by the CDN rather than the original.

SRI doesn't work with files that have been optimised by jsDelivr.

image

image

File: app/main/routes.py
Comment (anthonyhashemi) [reply app/main/routes.py:736]: for visibility: here https://iiif.io/guides/using_iiif_resources/ is an explanation of the need for a manifest and uri.

A manifest is the crucial resource for the iiif standard.

We are choosing to use universal viewer to render iiif resources in our webapp but by exposing this authenticated manifest url, we are giving our users the ability to go and render the manifest in another viewer / look at the raw details available in the manifest. Initially I was concerned that there could be some data here we would not want to expose but I believe that it is ok; the viewer is just a convenience; all the data is in the webapage already.

Would be worth sitting down and verifying this is the case and letting the team know that this endpoint will be available. Whether we want to advertise this endpoint is an interesting question.

File: app/main/routes.py
Comment (anthonyhashemi) [inline app/main/routes.py:736]: would be really good to add 1 or 2 high level tests for this new endpoint and then any edge casey ones fdor manifest creation at the generate_pdf_manifest and generate_image_manifest function levels. will act as a good example of what manifests should look like

File: app/templates/main/record-view.html
Comment (anthonyhashemi) [inline app/templates/main/record-view.html:1]: when is this file_type not iiif?

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py:122]: Set nonces to be used in script-src only 👍

File: app/templates/main/record-view.html
Comment (colinbowen) [reply app/templates/main/record-view.html:1]: File type will only be iiif for files that can be rendered by iiif & UV, see:

```
def get_file_details(file):
    """Retrieve file metadata and determine file type and extension."""
    file_metadata = get_file_metadata(file.FileId)
    file_extension = file.FileName.split(".")[-1].lower()

    if file_extension in ["pdf", "png", "jpg", "jpeg"]:
        file_type = "iiif"
    else:
        file_type = None

    return file_metadata, file_type, file_extension
```

If a render isn't available, we default to loading the record details view

Comment (colinbowen) [reply app/templates/main/record-view.html:1]: Just wanted to check that having this conditional here was ok?

File: app/static/init.uv.js
Comment (colinbowen) [reply app/static/init.uv.js:40]: This is just to ensure that UV is init correctly on either tab switch or the initial load when viewing the record page not for any form submissions. I think we spoke about this on the call but might be worth clarifying this one.

File: app/__init__.py
Comment (ahosgood) [inline app/__init__.py:73]: How easy are these going to be to maintain?

Is this relevant or have we tested it? [Firefox does not support hash-based CSP](https://bugzilla.mozilla.org/show_bug.cgi?id=1409200)

Comment (colinbowen) [reply app/__init__.py:73]: These hashes should be the same and shouldn't really change going forward as they are the hash of each style required by UV. We're also not using strict-dynamic in our CSP and I've done some testing in Firefox where the PDF is loaded correctly.

File: app/static/uv.min.css
Comment (colinbowen) [reply app/static/uv.min.css]: There's no difference to serving this file ourself as it was previously set to a specific version from the CDN. This changed during testing / development to try and debug the issues with the CSP. For the sake of simplicity both files are being pulled from the cdn.

**Comment by ltrasca** [comment]
The tabs container sticks out a little bit on each side compared to the elements above it (e.g. breadcrumbs and header elements), any way this could be fixed in this pr?

image

**Comment by colinbowen** [comment]
> The tabs container sticks out a little bit on each side compared to the elements above it (e.g. breadcrumbs and header elements), any way this could be fixed in this pr?
> 
> image

Moved the tabs-content to a govuk width container

File: app/__init__.py
Comment (ahosgood) [reply app/__init__.py:73]: So everything is good as long as we continue to use the same version of UV?

**Review by ltrasca — DISMISSED** [review]

File: app/__init__.py
Comment (colinbowen) [reply app/__init__.py:73]: Yes that's correct. So the CSP hashes in either script-src or style-src are to allow certain styles or inline scripts. See: https://content-security-policy.com/hash/

File: app/templates/main/record-view.html
Comment (anthonyhashemi) [reply app/templates/main/record-view.html:1]: sorry for the late response, I get it now. I assume @ltrasca expands this in his work to show the we don't support your document for render ticket?

File: app/static/init.uv.js
Comment (anthonyhashemi) [reply app/static/init.uv.js:40]: Its been a while since the call so yeah maybe wortth catching up again but my thoughts are, if we support the tab switching with non javascript then we'd need to do it via form submissions and thus wouldnt need javascript to do it.

Comment (colinbowen) [reply app/static/init.uv.js:40]: Check with Terry that UI side of changes is ok

File: app/static/init.uv.js
Comment (colinbowen) [reply app/static/init.uv.js:79]: Will ensure for full responsiveness after this has been merged. Ticket to track: https://national-archives.atlassian.net/browse/AYR-1198

File: app/main/routes.py
Comment (colinbowen) [reply app/main/routes.py:736]: Discussed with AYR, ok to leave this protected route in.

Comment (colinbowen) [reply app/main/routes.py:736]: Discussed on call on why this route is required for Universal Viewer. 
- UV requires a URL to access the manifest file from in JSON format
- Manifest data can be created and is behind a protected route

File: app/main/util/render_utils.py
Comment (anthonyhashemi) [inline app/main/util/render_utils.py]: I just realised that id should be the url of the manifest online, not the url of the file. https://iiif.io/api/cookbook/recipe/0001-mvm-image/

`The id property identifies this manifest with the URL at which it is available online.`

File: app/main/util/render_utils.py
Comment (anthonyhashemi) [inline app/main/util/render_utils.py]: same here

File: app/main/routes.py
Comment (colinbowen) [reply app/main/routes.py:736]: Added PDF generation & image generation manifest tests to ensure correct manifest is produced.

**Review by anthonyhashemi — DISMISSED** [review]
Spoke on call. Approving this PR to unblock accesibility testing but currently we cant allow a user without download permission to view a render. We will work on fixing this in a separate PR ASAP.

**Review by anthonyhashemi — APPROVED** [review]

**Pull request merged**
